contracts/FundDistribution.sol

```diff
@@ -0,0 +1,257 @@
+    require(to != address(0), "Invalid address");
+    _;
+  }
+
+  modifier onlyFunders() {
+    require(funders[msg.sender] || msg.sender == owner, "Only funders can call this function");
+    _;
+  }
+
+  modifier onlyDistributors() {
+    require(
+      distributors[msg.sender] || msg.sender == owner,
+      "Only distributors can call this function"
+    );
+    _;
+  }
+
+  function addFunder(address _funder) public onlyOwner {
+    funders[_funder] = true;
+  }
+
+  function addDistributor(address _distributor) public onlyOwner {
+    distributors[_distributor] = true;
+  }
+
+  receive() external payable onlyFunders onlyNonZeroAmount(msg.value) {
+    emit EthIsAdded(msg.sender, msg.value);
+  }
+
+  //don't emit an event as receive() already emitted it
+  function depositEth() public payable onlyFunders onlyNonZeroAmount(msg.value) {}
+
+  function depositEthAndApprove(address to) public payable {
+    depositEth();
+    setEthApprove(to, msg.value);
+  }
+
+  //transfer token to this contract first and then call this function
+  function addTokenAfterTransfer(address token) public onlyFunders onlyNonZeroAddress(token) {
+    require(!curTokens[token], "Token already added");
+    curTokens[token] = true;
+    tokens.push(token);
+    uint256 amount = IERC20(token).balanceOf(address(this));
+    require(amount > 0, "Token has no balance");
+    emit TokenIsAdded(msg.sender, token, amount);
+  }
+
+  // approve allowance first then call receiveToken to transfer token
+  function depositToken(address token, uint256 amount)
+    public
+    onlyFunders
+    onlyNonZeroAddress(token)
+    onlyNonZeroAmount(amount)
+  {
+    if (!curTokens[token]) {
+      tokens.push(token);
+      curTokens[token] = true;
+    }
+    IERC20 tokenContract = IERC20(token);
+    tokenContract.transferFrom(msg.sender, address(this), amount);
+    emit TokenIsAdded(msg.sender, token, amount);
+  }
+
+  function depositTokenAndApprove(
+    address token,
+    address to,
+    uint256 amount
+  ) external {
+    depositToken(token, amount);
+    setTokenApprove(to, token, amount);
+  }
+
+  //set the amount claimable to an address
+  function setEthApprove(address to, uint256 amount)
+    public
+    onlyDistributors
+    onlyNonZeroAddress(to)
+  {
+    ethAvailable[to] = amount;
+    emit EthApproveIsSet(to, amount);
+  }
+
+  function setEthApproveMultiple(address[] calldata tos, uint256[] calldata amounts) external {
+    require(tos.length == amounts.length, "Invalid array lengths");
+    for (uint256 i = 0; i < tos.length; i++) {
+      setEthApprove(tos[i], amounts[i]);
+    }
+  }
+
+  //set the token amount claimable to an address
+  function setTokenApprove(
+    address to,
+    address token,
+    uint256 amount
+  ) public onlyDistributors onlyNonZeroAddress(to) {
+    require(curTokens[token], "Token is not added");
+    tokenAvailable[to][token] = amount;
+    emit TokenApproveIsSet(to, token, amount);
+  }
+
+  function setTokenApproveMultiple(
```

Comment:
I think this function will still work if you pass in an array with one item, so there might not be a need for the `setTokenApproval` function for single token addresses right? Unless it saves gas or something

Reply:
Yep, you're right. It's not gonna save much gas, but I just ensure that people will not have any problem if they forgot to pass in an array for just a single token =)))